Implementação da HTTP REST API

joaojvds opened the pull request:
--Implementação das rotas
--Implementação dos testes da API HTTP

GCastilho committed:
Reestruturado a pasta router

-- A pasta router agora é a pasta base do web server inteiro, não apenas do roteador
-- O módulo 'server' agora é o index da server

GCastilho committed:
Adicionado handler para subdomínios

-- Adicionado um handler para subdomínios, requests para um subdomínio inexistente retornam 404
-- Adicionado subdomínio 'api' para requests à httpApi

GCastilho committed:
Módulo do servidor agora está em TypeScript

-- O módulo principal do webserver foi convertido para TypeScript
-- Os módulos dos subdomínios foram adaptados para funcionarem com o módulo principal em TypeScript

GCastilho committed:
Adicionado lista de testes

-- Adicionado lista dos testes que serão implementados na HTTP API

GCastilho committed:
Adicionado teste para informações sobre a API

-- Adicionado teste da entrypoint da v1 da API, que retorna informações sobre a API
-- Adicionado um teste para a root da api, que retorna uma lista das versões suportadas

joaojvds committed:
Criado rotas para as chamadas http

--O httpApi tem todas as rotas presentes nos testes

GCastilho committed:
Adicionado teste para /currencies

-- Adicionado um teste para /v1/currencies
-- Refatorado os testes para usar o expect ao invés do assertion do supertest
-- Modificado resposta do /v1 para remover informações duplicadas e ficar mais claro

GCastilho committed:
Adicionado teste para user

-- Adicionado teste para o entrypoint da user
-- O request de wirhdraw será feito usando um POST em transactions
-- O endpoint da transactions agora fica sob a user

GCastilho committed:
Adicionado testes para 401

-- Adicionado os testes para 401 nos endpoints que os utilizam
-- Adicionado um 'before' na /user para criar um usuário e fazer login urilizando o /login

joaojvds committed:
implementado path currencies

GCastilho committed:
Resolvido bug na comparação de objetos

-- Resolvido um bug onde objetos retornados não estavam sendo comparados corretamente nos testes
-- Minor changes

GCastilho committed:
Implementado teste para /accounts

-- Accounts de um usuário são manualmente adicionadas no before da user
-- Adicionado um teste para accounts de um usuário
-- Bugfix do último commit

GCastilho committed:
Adicionado teste para /balances

-- Adicionado testes para /balances
-- Adicionado parâmetro opcional na getBalance para retornar os saldos como strings
-- Bugfix de um teste

GCastilho committed:
Adicionado teste para /transactions

-- Adicionado um teste para o retorno do /transactions
-- Adicionado teste para /transactions/:opid que não pertence a sua account
-- Adicionado transações dummy no 'before' da transactions para ter transações para verificar

GCastilho committed:
Adicionado teste para pegar tx por opid

-- Adicionado um teste para /transactions/:opid que espera um retorno equivalente ao get_tx_info
-- O before do transaction agora executa os requests alternando entre as currencies, ao invés de uma de cada vez
-- Bugfix

GCastilho committed:
Adicionado teste para withdraw

-- Adicionado um teste para withdraw
-- Adicionado teste para withdraw sem fundos suficientes
-- Adicionado overload de métodos da getBalance da User para documentar corretamente seu retorno
-- O teste para info de transação por opid for expandido para testar todas as transações do database
-- Bugfix no teste de not authorized das transactions

joaojvds committed:
adicionado check do cookie

--Envia Not Authorized se não houver sessionId ou ele não for valido
--minor changes

GCastilho commented:
Vc está fazendo os check do cookie em todos os handlers, pq não faz um único em um middleware antes deles?

joaojvds committed:
Adicionado middleware no user.ts

--Colocado check do cookie em um middleware

GCastilho committed:
Separado libs de types

-- Libs que simplesmente expandiam types do typescript foram movidas para uma pasta @types, a pasta libs será apenas para módulos que expandem funções de algo
-- Adicionado type para permitir referência à user no request do express

GCastilho committed:
Limpado check do 'not authorized'

-- Checks da /user por 'not authorized' foram substituídos de um objeto de erro individual por teste por uma referência a um único objeto de erro

GCastilho committed:
Adicionado teste para filtro de transações por currency

-- Adicionado teste para filtro de transações por currency
-- Adicionado teste para pular primeiras n transações
-- Adicionado teste para GET de transações quando não se tem transações na conta
-- Modificado teste do /user para retornar corretamente informações sobre o filtro de transações

GCastilho committed:
Adicionado testes para not found

-- Adicionado testes para not found na /v1 e na /v1/user
-- Alguns bugfixes

GCastilho committed:
Corrigido um bug nos testes de cookie

-- Corrigido um bug nos testes de cookie

joaojvds committed:
Work in progress

--Implementando /Account e /balances (precisa ser testado)
--minor changes

GCastilho committed:
Resolvido um bug no script de compilação

-- Resolvido um bug no script de compilação que impedia a compilação em sistemas linux

joaojvds committed:
Trabalhando no handle da transactions

--implementando handle da transactions e da transactions/opid
--Minor changes

GCastilho committed:
Resolvido alguns bugs nos testes

GCastilho committed:
Corrigido um bug na manipulação do opid

-- Corrigido um bug onde estava-se tentando acessar o método toHexString do opid vindo da api que já é uma string

joaojvds committed:
Implementado requests withdraw

--Ao da post na transactions ele realiza o request de withdraw

joaojvds committed:
Minor fix

--Agora o sessionId é enviado nos tests '/' e no '/notFoundPath'

GCastilho committed:
Corrigido checks de propriedades de transações

-- Corrigido como é feito os checks por propriedades de uma transação
-- Corrigido um bug na ordem das transações esperadas ao requisitar várias
-- Transações salvas no db agora tem txid e confirmations

joaojvds committed:
Implementado request de varias transações

--Função agora pega as dez ultimas transações do usuário, podendo escolher pegar transações de uma moeda em especifica ou então pular um determinado numero de transações.
--Consertado bug na comparação dos opid

joaojvds committed:
Implementado requests na account e na balance

--Consertado bug que fazia ele levar mais de 2000ms para responder

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
 router.get('/transactions/:opid', async (req, res) => {
 	try {
-		const transactions = await Transaction.find({ _id: req.params.opid })
+		const transactions = await Transaction.findOne({ _id: req.params.opid })
```
tem um método que chama "findById", que é exatamente o que vc precisa. Usa ele

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
-		})
+		if (err === 'NotFound') {
+			res.status(404).send({
+				error: err,
```
Eu acho melhor colocar o "NotFound" daqui explicitamente. Pode dar problema se mudar o erro lá em cima além de ficar mais "difuso", pq não fica claro à primeira vista o que é retornado

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
+			})
+		} else {
+			res.status(401).send({
+				error: err,
```
O mesmo aqui, coloca explicito

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
 
-//TODO: consertar o erro "expected undefined to be an object"
+/**
+ * Pega multiplas transações feitas pelo usuario
```
Troca por "Retorna uma lista de transações do usuário"

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
+ */
 router.get('/transactions', async (req, res) => {
 	/** Indica o numero de transações que sera puladas */
 	const skip: number = +req.query.skip || 0
```
Se skip for não-numérico isso vai retornar um NaN, que vai fazer skip ser NaN e não 0, que ctz q vai dar algum problema em algum lugar

joaojvds replied:
Ele vai retornar 0 mesmo se for NaN, eu testei só para ter certeza inclusive testei no navegador e retornou 0.
![image](https://user-images.githubusercontent.com/28350465/77344284-b5314c80-6d11-11ea-836c-770f33a3d276.png)

GCastilho replied:
Ah, então excelente

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
 		req.query.currency ? {
 			user: req.user?.id,
 			currency: req.query.currency
 		} : { user: req.user?.id })
```
Isso daqui tá uma zona, mds eu nem sei o que isso retorna

Bota esse ternário numa variável auxiliar, pelo amor de deus

Vc tbm tem que checar se "req.param.currency" é uma das SuportedCurrencies da currencyApi antes de continuar

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
-
-	res.send(transactions.slice(skip, skip + 10))
+		.sort({ timestamp: -1 }))
+		.slice(skip, skip + 10)
```
Usar o slice vai fazer TODAS as transações do usuário serem enviadas pra memória depois o array ser cortado. Agora imagina q são 1k de usuários com 50 transações. PARECE q vai dar um pico de memória aqui

Tenta ver se o mongo tem um filtro pra retornar apenas as que precisa. **Se não tiver** usa um cursor

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -67,42 +67,70 @@ router.get('/info', async (_req, res) => {
+			confirmations:	transaction.confirmations,
+			timestamp:		transaction.timestamp.getTime()
+		})
+	})
```
usar for...of ao invés de forEach pq forEach é mais lento e gasta mais recursos

GCastilho reviewed (commented):
Legal que vc descobriu o findOne, tava preocupado que vc não fosse descobrir ele :v

joaojvds committed:
Resolvido alguns problemas

--Agora ele checa se as propriedades tem o type correto antes de enviar para a currencyApi
--Trocado findOne por findById
--Colocado um check para ver se a currency enviada é uma currency suportada
--Trocado método .slice por um método do Mongoose
--Minor changes

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+
+const router = express.Router()
+
+/** Parsers */
```
troca esse comment de JSDoc para comentário normal, com //

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+
+/** Parsers */
+router.use(cookieParser()) // for parsing cookie
+router.use(express.json()) // for parsing application/json
```
Não precisa desses comentário explicando pra que é o parser, é auto-explicativo. Pode retirar

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+ */
+router.get('/transactions/:opid', async (req, res) => {
+	try {
+		const transactions = await Transaction.findById(req.params.opid)
```
O que acontece se o parâmetro opid for inválido ou nulo?

joaojvds replied:
Se ele for invalido ele não vai encontrar a transação o que vai cair no if e disparar um not found, agora quando você diz nulo, você quer dizer quando ele não envia o opid?
Se for isso, então ele não vai cair ai no '/transactions/:opid' e sim no '/transactions'

GCastilho replied:
Hmmmm faz sentido

E se a string do opid for de um tamanho diferente do tamanho que um objectId tem (maior ou menor)? Não dá nenhum erro?

joaojvds replied:
Se a string for invalida seja por tamanho diferente ou algo problema, transactions vai retornar um undefined o que vai ser barrado no if dando throw 'notFound'

GCastilho replied:
Hmmm certo. Se vc **testou** e realmente é isso então ok

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+		const transactions = await Transaction.findById(req.params.opid)
+		if (!transactions) throw 'NotFound'
+		//console.error(`opid=${transactions._id} user=${transactions.user.toHexString()}`)
+		/** Checa se o usuario da transação é o mesmo que esta logado */
```
Coloca esse comment e o de baixo como normais, usando //

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+		if (transactions.user.toHexString() !== req.user?.id.toHexString()) throw 'NotAuthorized'
+		/** Coloca as transações o formato certo */
+		const tx = {
+			opid:			transactions._id,
```
Coloca toHexString nesse id

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+	const transactions = await Transaction.find(query, null,{
+		sort : { timestamp: -1 },
+		limit: 10,
+		skip: skip
```
Qdo a variável tem o mesmo nome do parâmetro do objeto que vc vai colocar ela como propriedade, pode colocar só o nome da propriedade, ou seja só "skip" que tem o mesmo resultado

GCastilho commented on src/server/api/v1/user.ts:
```diff
@@ -0,0 +1,266 @@
+		if (!currency ||
+			!req.user ||
+			typeof req.body.destination !== 'string' ||
+			isNaN(+req.body.amount)) throw 'BadRequest'
```
Bota o operador desse OR no começo da linha, aí eles ficam alinhados

GCastilho reviewed (commented):
Quero anotar que comentários JSDoc são para declaração de coisas, tipo variáveis, funções e etc. Não pra o que um bloco de código faz

Então "Faz x, y, z" usa "//" e comentário normal, se extender múltiplas linhas aí usa /* comment */ e "calcula a raiz quadrada" como comentário antes de uma função usa o JSDoc

joaojvds committed:
Mudanças nos comentários

--Trocado alguns comentários do JSDoc para comentários normais

GCastilho committed:
Resolvido um bug no teste de transação específica

-- Resolvido um bug onde o teste de informações de uma transação específica estava requisitando txs que não era do usuário de testes

GCastilho committed:
Removido documentação dos requests da API

-- Os requests de documentação foram retirados da API, faz mto mais sentido essas infomações estarem numa docs do que na api
-- Teste do versions da root da api foi substituído por um array 'entries', com os subpaths da api ao invés das versões dela

joaojvds committed:
Minor changes

--Minor changes

GCastilho committed:
Merge branch 'master' into httpApi

GCastilho committed:
Corrigido bug onde o amount esperado de um teste não estava levando o fee em consideração

joaojvds committed:
Corrigido bug no endpoint /currencies

GCastilho committed:
Minor fixes

-- Corrigido alguns erros de formatação na user

GCastilho committed:
Substituído for..of da /transactions

-- Substituido for..of da /transactions por um map, já que os testes de performance mostraram que a diferença é aceitável

GCastilho committed:
minor fix

GCastilho reviewed (approved):
Aprovado com louvor!
